src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,199 @@
+        }
+    }
+
+    private void areIndexInRange(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + " doesn't exist");
+        }
+    }
+
+    private int findIndex(T object) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            T currentObject = node.item;
+            if (areItemsEqual(currentObject, object)) {
+                return i;
+            } else {
+                node = node.next;
+            }
+        }
+        return FALSE_INDEX;
+    }
+
+    Node<T> node(int index) {
+        areIndexInRange(index);
+        Node<T> wantedNode;
+        if (index < (size >> 1)) {
+            wantedNode = first;
+            for (int i = 0; i < index; i++) {
+                wantedNode = wantedNode.next;
+            }
+        } else {
+            wantedNode = last;
+            for (int i = size - 1; i > index; i--) {
+                wantedNode = wantedNode.prev;
+            }
+        }
+        return wantedNode;
+    }
+
+    private void linkReassignmentForAdd(T value, int index) {
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+        Node<T> indexNode = node(index);
+        if (indexNode.prev == null) {
+            linkFirst(value, indexNode);
+        } else {
+            linkBefore(value, indexNode);
+        }
+    }
+
+    private void linkReassignmentForRemove(int index) {
+        Node<T> node = node(index);
+        if (index == 0) {
+            unlinkFirst(node);
+        } else if (index == size - 1) {
+            unlinkLast(node);
+        } else {
+            unlinkBefore(node);
+        }
+        size--;
+    }
+
+    private void unlinkFirst(Node<T> node) {
+        Node<T> nextNode = node.next;
+        if (nextNode != null) {
+            nextNode.prev = null;
+            node.next = null;
+            first = nextNode;
+        } else {
+            first = null;
+            last = null;
+        }
+    }
+
+    private void unlinkLast(Node<T> node) {
+        Node<T> previousNode = node.prev;
+        previousNode.next = null;
+        node.prev = null;
+        last = previousNode;
+    }
+
+    private void unlinkBefore(Node<T> node) {
+        Node<T> previousNode = node.prev;
+        Node<T> nextNode = node.next;
+        previousNode.next = nextNode;
+        nextNode.prev = previousNode;
+    }
+
+    private void linkFirst(T value, Node<T> node) {
+        Node<T> newNode = new Node<>(null, value, node);
+        node.prev = newNode;
+        first = newNode;
+        size++;
+    }
+
+    private void linkLast(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        if (last == null) {
+            first = newNode;
+        } else {
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> indexNode) {
```

Comment:
method naming suggestion
```suggestion
    private void linkMiddle(T value, Node<T> indexNode) {
```